src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/AbstractSyntaxFactsService.cs

```diff
@@ -488,12 +488,47 @@ public bool SpansPreprocessorDirective(IEnumerable<SyntaxToken> tokens)
         private bool SpansPreprocessorDirective(SyntaxTriviaList list)
             => list.Any(t => IsPreprocessorDirective(t));
 
-        public bool IsInHeader(SyntaxNode node, SyntaxNode ownerOfHeader, SyntaxNodeOrToken lastTokenOrNodeOfHeader)
+        public bool IsOnHeader(int position, SyntaxNode ownerOfHeader, SyntaxNodeOrToken lastTokenOrNodeOfHeader)
         {
             var start = GetStartOfNodeExcludingAttributes(ownerOfHeader);
             var end = lastTokenOrNodeOfHeader.FullSpan.End;
 
-            return node.Span.Start >= start && node.Span.End <= end;
+            return start <= position && position <= end;
+        }
+
+        public bool IsInHeader(int position, SyntaxNode ownerOfHeader, SyntaxNodeOrToken lastTokenOrNodeOfHeader, ImmutableArray<SyntaxNode> holes)
+        {
+            var inHeader = IsOnHeader(position, ownerOfHeader, lastTokenOrNodeOfHeader);
+            if (!inHeader)
+            {
+                return false;
+            }
+
+            // Holes are exclusive: to return false it needs to be _inside_ a hole not only on the edge
+            // -> to be consistent with other 'being on the edge' of Tokens/Nodes.
+            if (holes.Any(h => (h.Span.Start < position && position < h.Span.End)))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Tries to get an ancestor of a Token on current position or of Token directly to left:
+        /// e.g.: tokenWithWantedAncestor[||]tokenWithoutWantedAncestor
+        /// </summary>
+        protected static TNode TryGetAncestorForLocation<TNode>(int position, SyntaxNode root) where TNode : SyntaxNode
+        {
+            var token = root.FindToken(position);
+            var node = token.GetAncestor<TNode>();
+            if (node == null && token.FullSpan.Start == position)
+            {
+                token = token.GetPreviousToken();
+                node = token.GetAncestor<TNode>();
+            }
+
+            return node;
```

Comment:
nit:

```c#
if (node != null)
    return node;

...

return token.GetPreviousToken().GetAncestor<TNode>();
```

I find the need to track data-flow over variables particularly annoying.  Because now i have to say "under what circumstances will this decisoin be overrules/overassigned?".  If we can just return the values when we know they're good, it's much simpler for my old and addled brain.